### PR TITLE
[Usage] Create VPC billing

### PR DESCRIPTION
**Pull request by SadiJr**
### Description

The usage of VPCs are not captured and saved by the Usage plugin, which makes it impossible to charge via the Quota plugin. The Usage plugin has extended, to save the VPC events in database. When a VPC is created, the event `EVENT_VPC_CREATE` is used; when a VPC is removed, the event `EVENT_VPC_DELETE` is used. Also, this new type has the attributes `value.id` and `value.name` injected in the Quota plugin.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?
It was tested in a local lab:

1. I created and destroyed VPCs via UI;
2. Before, ACS did not save any VPC event in Usage;
3. Now, ACS saves these events in Usage.

### CODE REVIEW

**Comment by DaanHoogland** [comment]
Wouw, I would not have guessed this omission was there, good work @SadiJr

File: engine/schema/src/main/java/com/cloud/usage/dao/UsageVpcDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            SearchCriteria<UsageVpcVO> sc = this.createSearchCriteria();
+            sc.addAnd("vpcId", SearchCriteria.Op.EQ, usage.getVpcId());
+            sc.addAnd("created", SearchCriteria.Op.EQ, usage.getCreated());
+            UsageVpcVO vo = findOneBy(sc);
+            if (vo != null) {
+                vo.setRemoved(usage.getRemoved());
+                update(vo.getId(), vo);
+            }
+        } catch (final Exception e) {
+            LOGGER.error(String.format("Error updating backup metrics due to [%s].", e.getMessage()), e);
```
Comment (stephankruggg) [inline engine/schema/src/main/java/com/cloud/usage/dao/UsageVpcDaoImpl.java:55]:
```suggestion
            LOGGER.error(String.format("Error updating usage of VPC due to [%s].", e.getMessage()), e);
```

File: engine/schema/src/main/java/com/cloud/usage/dao/UsageVpcDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+                    createdDate = DateUtil.parseDateString(s_gmtTimeZone, createdTS);
+                }
+                if (removedTS != null) {
+                    removedDate = DateUtil.parseDateString(s_gmtTimeZone, removedTS);
+                }
+                usageRecords.add(new UsageVpcVO(id, vpcId, zoneId, acctId, domId, stateTS, createdDate, removedDate));
+            }
+        } catch (Exception e) {
+            txn.rollback();
+            LOGGER.warn("Error getting VM backup usage records", e);
```
Comment (stephankruggg) [inline engine/schema/src/main/java/com/cloud/usage/dao/UsageVpcDaoImpl.java:124]:
```suggestion
            LOGGER.warn("Error getting VPC usage records", e);
```

File: engine/schema/src/main/java/com/cloud/usage/dao/UsageVpcDaoImpl.java (hunk 1 above)
Comment (SadiJr) [reply engine/schema/src/main/java/com/cloud/usage/dao/UsageVpcDaoImpl.java]: Done, thanks.

File: engine/schema/src/main/java/com/cloud/usage/dao/UsageVpcDaoImpl.java (hunk 2 above)
Comment (SadiJr) [reply engine/schema/src/main/java/com/cloud/usage/dao/UsageVpcDaoImpl.java]: Done, thanks.

**Review by stephankruggg — APPROVED** [review]
LGTM based on manual tests

**Comment by kiranchavala** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@SadiJr Could you please resolve the merge conflicts

**Comment by DaanHoogland** [comment]
@SadiJr we can target this for 4.18.1 or for 4.19 . In any case th upgrade script needs to be altered (now using 4.17.2 to 4.18.0)

**Comment by SadiJr** [comment]
@DaanHoogland done, thanks for your suggestion.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by SadiJr** [comment]
> @SadiJr can you look at these: https://github.com/apache/cloudstack/actions/runs/6190568015/job/16883290403?pr=7235#step:6:9441 ?

Done, thanks.

File: api/src/main/java/com/cloud/event/EventTypes.java [hunk 3]
```diff
@@ -1172,4 +1172,8 @@ public static Class getEntityClassForEvent(String eventName) {
 
         return null;
     }
+
+    public static boolean isVpcEvent(String eventType) {
+        return EventTypes.EVENT_VPC_CREATE.equals(eventType) || EventTypes.EVENT_VPC_DELETE.equals(eventType);
```
Comment (DaanHoogland) [inline api/src/main/java/com/cloud/event/EventTypes.java:1177]: not for update?

File: engine/schema/src/main/resources/META-INF/db/schema-41800to41810.sql [hunk 4]
```diff
@@ -47,3 +47,17 @@ WHERE so.default_use = 1 AND so.vm_type IN ('domainrouter', 'secondarystoragevm'
...
+  `zone_id` bigint(20) unsigned NOT NULL,
+  `account_id` bigint(20) unsigned NOT NULL,
+  `domain_id` bigint(20) unsigned NOT NULL,
+  `state` varchar(100) DEFAULT NULL,
+  `created` datetime NOT NULL,
+  `removed` datetime DEFAULT NULL,
+  PRIMARY KEY (`id`)
+) ENGINE=InnoDB CHARSET=utf8;
+
+ALTER TABLE `cloud_usage`.`cloud_usage` ADD state VARCHAR(100) DEFAULT NULL;
```
Comment (DaanHoogland) [inline engine/schema/src/main/resources/META-INF/db/schema-41800to41810.sql:63]: needs to be moved to a newer upgrade file

**Comment by DaanHoogland** [comment]
@stephankruggg , I am assuming you executed the tests as per the description?

> LGTM based on manual tests

**Comment by stephankruggg** [comment]
> @stephankruggg , I am assuming you executed the tests as per the description?
> 
> > LGTM based on manual tests

Yes

File: engine/schema/src/main/resources/META-INF/db/schema-41800to41810.sql (hunk 4 above)
Comment (DaanHoogland) [reply engine/schema/src/main/resources/META-INF/db/schema-41800to41810.sql:63]: @SadiJr this really nees addressing.

**Comment by DaanHoogland** [comment]
@SadiJr @stephankruggg as there is work to do here, I am moving it to unplanned (not sure who else to cc)

**Comment by SadiJr** [comment]
@DaanHoogland Sorry for the delay, I will check the suggestions and work on this PR

File: api/src/main/java/com/cloud/event/EventTypes.java (hunk 3 above)
Comment (SadiJr) [reply api/src/main/java/com/cloud/event/EventTypes.java:1210]: @DaanHoogland from my point of view, as the VPC update API only changes metadata such as name, MTU, etc., I see no need to handle the update event here.

File: engine/schema/src/main/resources/META-INF/db/schema-41800to41810.sql (hunk 4 above)
Comment (SadiJr) [reply engine/schema/src/main/resources/META-INF/db/schema-41800to41810.sql]: @DaanHoogland sorry for the delay, and done, thanks for your review.

**Comment by BryanMLima** [comment]
@SadiJr, the lint is failing due to the end of the schema files. You need to add an extra line at the end.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Review by BryanMLima — APPROVED** [review]
CLGTM

**Review by DaanHoogland — APPROVED** [review]
clgtm, but needs testing

**Comment by BryanMLima** [comment]
As @SadiJr is focusing in other tasks, I will be working on this PR.

**Comment by BryanMLima** [comment]
@blueorangutan package

**Comment by BryanMLima** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
> [SF] Trillian test result (tid-9407) Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7 Total time taken: 49287 seconds Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr7235-t9407-kvm-centos7.zip Smoke tests completed. 128 look OK, 1 have errors, 0 did not run Only failed and skipped tests results shown below:
> Test 	Result 	Time (s) 	Test File
> test_02_list_cpvm_vm 	`Failure` 	0.04 	test_ssvm.py
> test_04_cpvm_internals 	`Failure` 	0.03 	test_ssvm.py

looks not related but will retest just to be sure

**Comment by DaanHoogland** [comment]
@blueorangutan test alma9 kvm-alma9 keepEnv

**Comment by BryanMLima** [comment]
@blueorangutan package

**Comment by BryanMLima** [comment]
@DaanHoogland  @sureshanaparti, do you guys think this PR can go to the `4.19.1.0` release?

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by BryanMLima** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@stephankruggg did you test this?

**Comment by stephankruggg** [comment]
> @stephankruggg did you test this?

Yes, prior to approving this PR I ran the same tests as @SadiJr, as far as I can remember.

**Pull request merged**